### ISSUE TITLE
InaccessibleMethod emitted even when constructor isn't misused

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#483]
I think it was introduced by 684eb7923b6c16274ec596bd850ef7259804ebd4 , but didn't investigate deeply

There are two types of valid code that would be affected:

1. Code using https://github.com/marc-mabe/php-enum
2. Classes which add private constructors to ensure they can't be instantiated (e.g. collections of static methods)

Example code:

```php
<?php

class Enum {
    private function __construct() {
        return new self();
    }
    /** @var static[] */
    private static $cache = [];

    /** @return static[] */
    public static function instance() {
        $value = self::$cache[static::class] ?? null;
        if ($value === null) {
            $value = new self();
            self::$cache[static::class] = $value;
        }
        return $value;
    }
}

class EnumSubclass extends Enum {}  // Observed: InaccessibleMethod. Expected: Nothing

class OtherEnum extends EnumSubclass {  // Observed: InaccessibleMethod
}

// $value = OtherEnum::instance();
```

**Comment by TysonAndre** [comment]
This is already covered by warnings about `new EnumSubclass` and `new OtherEnum` should be warned about. (`InaccessibleMethod - src/example.php:26:1 - Cannot access private method Enum::__construct from context` if `new OtherEnum()` is added)

Or create a new issue type entirely so that this can be suppressed entirely without suppressing real issues.

**Comment by muglug** [comment]
> One solution to the false positive in the example is to emit a distinct issue type (e.g. `InaccessibleInheritedConstructor`) so that this can be suppressed entirely without suppressing real issues.

Great idea

**Comment by muglug** [comment]
Actually I'll just revert this, as it's not really fixing anything.

**Comment by muglug** [comment]
Sorry for the trouble!
